README.md
# pynock

![Licence](https://img.shields.io/github/license/DerwenAI/pynock)
![Repo size](https://img.shields.io/github/repo-size/DerwenAI/pynock)
![GitHub commit activity](https://img.shields.io/github/commit-activity/w/DerwenAI/pynock?style=plastic)
[![Checked with mypy](http://www.mypy-lang.org/static/mypy_badge.svg)](http://mypy-lang.org/)
![CI](https://github.com/DerwenAI/pynock/workflows/CI/badge.svg)
![downloads](https://img.shields.io/pypi/dm/pynock)
![sponsor](https://img.shields.io/github/sponsors/ceteri)

The following describes a proposed standard `NOCK` for a Parquet
format that supports efficient distributed serialization of multiple
kinds of graph technologies.

This library `pynock` provides Examples for working with low-level
Parquet read/write efficiently in Python.

Our intent is to serialize graphs in a way which aligns the data
representations required for popular graph technologies and related
data sources:

  * semantic graphs (e.g., W3C formats RDF, TTL, JSON-LD, etc.)
  * labeled property graphs (e.g., openCypher)
  * probabilistic graphs (e.g., PSL)
  * spreadsheet import/export (e.g., CSV)
  * dataframes (e.g., Pandas, Dask, Spark, etc.)
  * edge lists (e.g., NetworkX, cuGraph, etc.)

This approach also efficient distributed partitions based on Parquet,
which can scale on a cluster to very large (+1 T node) graphs.

For details about the proposed format in Parquet files, see the
[`FORMAT.md`](https://github.com/DerwenAI/pynock/blob/main/FORMAT.md)
file.

If you have questions, suggestions, or bug reports, please open
[an issue](https://github.com/DerwenAI/pynock/issues)
on our public GitHub repo.


## Caveats

Note that the `pynock` library does not provide any support for graph
computation or querying, merely for manipulating and validating
serialization formats.

Our intent is to provide examples where others from the broader open
source developer community can help troubleshoot edge cases in
Parquet.


## Dependencies

This code has been tested and validated using Python 3.8, and we make
no guarantees regarding correct behaviors on other versions.

The Parquet file formats depend on Arrow 5.0.x or later.

For the Python dependencies, the library versioning info is listed in the
[`requirements.txt`](https://github.com/DerwenAI/pynock/blob/main/requirements.txt)
file.


## Set up

To install via PIP:

```
python3 -m pip install -U pynock
```

To set up this library locally:

```
python3 -m venv venv
source venv/bin/activate

python3 -m pip install -U pip wheel
python3 -m pip install -r requirements.txt
```

## Usage via CLI

To run examples from CLI:

```
python3 cli.py load-parq --file dat/recipes.parq --debug
```

```
python3 cli.py load-rdf --file dat/tiny.ttl --save-csv foo.csv
```

For further information:

```
python3 cli.py --help
```

## Usage programmatically in Python

To construct a partition file programmatically, see the 
[`examples`](https://github.com/DerwenAI/pynock/blob/main/examples)
for Jupyter notebooks with sample code and debugging.


## Background

For more details about using Arrow and Parquet see:

["Apache Arrow homepage"](https://arrow.apache.org/)

["Finer-grained Reading and Writing"](https://arrow.apache.org/docs/python/parquet.html#finer-grained-reading-and-writing)

["Apache Arrow: Read DataFrame With Zero Memory"](https://towardsdatascience.com/apache-arrow-read-dataframe-with-zero-memory-69634092b1a)  
Dejan Simic  
_Towards Data Science_ (2020-06-25)


## Why the name?

A `nock` is the English word for the end of an arrow opposite its point.

If you must have an acronym, the proposed standard `NOCK` stands for
**N**etwork **O**bjects for **C**onsistent **K**nowledge.

Also, the library name had minimal namespace collisions on GitHub and
PyPi :)


## Developer updates

To set up the build environment locally, also run:
```
python3 -m pip install -U pip setuptools wheel
python3 -m pip install -r requirements-dev.txt
```

Note that we require the use of [`pre-commit` hooks](https://pre-commit.com/)
and to configure that locally:

```
pre-commit install
git config --local core.hooksPath .git/hooks/
```


## Package releases

First, verify that `setup.py` will run correctly for the package
release process:

```
python3 -m pip install -e .
python3 -m pytest -rx tests/
python3 -m pip uninstall pynock
```

Next, update the semantic version number in `setup.py` and create a
release on GitHub, and make sure to update the local repo:

```
git stash
git checkout main
git pull
```

Make sure that you have set up your 2FA authentication for generating
an API token on PyPi: <https://pypi.org/manage/account/token/>

Then run our PyPi push script:

```
./bin/push_pypi.sh
```


## Star History

[![Star History Chart](https://api.star-history.com/svg?repos=derwenai/pynock&type=Date)](https://star-history.com/#derwenai/pynock&Date)


changelog.txt
# `pynock` changelog

## 1.2.1

2022-10-11

  * update dependencies, kudos @Mec-iS
  * improve docs about `edge_id` requirements, kudos @Mec-iS


## 1.2.0

2022-10-07

  * migrate sample code to `examples/` Jupyter notebooks
  * create an example of partitioning based on `NetworkX` "Karate Club"
  * add convenience methods `Partition.find_or_create_node()` and `Partition.create_edge()`


## 1.1.1

2022-10-06

  * add input validation and warnings
  * resolve conflicts with `pandas` for missing values
  * consistent parsing and generation of RDF
  * create CI pipeline using docker and GH Actions


## 1.0.1

2022-10-03

  * propose open standard


## 1.0.0

2022-10-02

  * first distribution on PyPi
  * initial check-in


tests/test_tiny.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Unit test coverage:

  * construct a partition programmatically
  * compare with a reference CSV file
"""

import tempfile

from icecream import ic
import cloudpathlib
import pytest

from pynock import Partition, Node, Edge


def test_tiny ():
    try:
        load_csv: str = "dat/tiny.csv"
        tmp_obs = tempfile.NamedTemporaryFile(mode="w+b", delete=True)

        # construct a Partition
        part: Partition = Partition(
            part_id = 0,
        )

        # lookup/create the src node for the recipe
        # NB: this node has properties, which RDF cannot query
        src_name: str = "https://www.food.com/recipe/327593"
        src_node: Node = part.find_or_create_node(src_name)

        src_node.is_rdf = True
        src_node.label_set = set(["Recipe"])
        src_node.prop_map = {
            "minutes": 8,
            "name": "anytime crepes",
        }

        # lookup/create a dst node for the "Egg" ingredient
        dst_name: str = "http://purl.org/heals/ingredient/ChickenEgg"
        dst_node: Node = part.find_or_create_node(dst_name)

        dst_node.is_rdf = True
        dst_node.label_set = set(["Ingredient"])

        # define an edge connecting src => dst for this ingredient
        part.create_edge(
            src_node,
            "http://purl.org/heals/food/uses_ingredient",
            dst_node,
        )

        # define a dst node for the "Milk" ingredient
        dst_name = "http://purl.org/heals/ingredient/CowMilk"
        dst_node = part.find_or_create_node(dst_name)

        dst_node.is_rdf = True
        dst_node.label_set = set(["Ingredient"])

        # define an edge connecting src => dst for this ingredient
        part.create_edge(
            src_node,
            "http://purl.org/heals/food/uses_ingredient",
            dst_node,
        )

        # define a dst node for the "Flour" ingredient
        # NB: this node has properties, which RDF cannot query
        dst_name = "http://purl.org/heals/ingredient/WholeWheatFlour"
        dst_node = part.find_or_create_node(dst_name)

        dst_node.is_rdf = True
        dst_node.label_set = set(["Ingredient"])
        dst_node.prop_map = {
            "vegan": True,
        }

        # define an edge connecting src => dst for this ingredient
        part.create_edge(
            src_node,
            "http://purl.org/heals/food/uses_ingredient",
            dst_node,
        )

        # define a dst node for the "wtm:Recipe" parent
        dst_name = "http://purl.org/heals/food/Recipe"
        dst_node = part.find_or_create_node(dst_name)

        dst_node.is_rdf = True
        dst_node.label_set = set(["top_level"])

        # define an edge connecting src => dst for this inheritance
        part.create_edge(
            src_node,
            "http://www.w3.org/1999/02/22-rdf-syntax-ns#type",
            dst_node,
        )

        # write the partition as a CSV file
        part.save_file_csv(
            cloudpathlib.AnyPath(tmp_obs.name),
            encoding = "utf-8",
            sort = True,
        )

        # compare the respective texts
        obs_text: str = cloudpathlib.AnyPath(tmp_obs.name).read_text()
        exp_text: str = cloudpathlib.AnyPath(load_csv).read_text()

        assert exp_text == obs_text

    except Exception as ex:
        ic(ex)

    finally:
        tmp_obs.close()


tests/test_pandas.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Unit test coverage:

Pandas handling of missing values:

  * read a Parquet file
  * read a CSV file
"""

from icecream import ic
import cloudpathlib
import pandas as pd
import pytest

from pynock import Partition


def test_pandas ():
    df_csv = pd.read_csv(
        cloudpathlib.AnyPath("dat/tiny.csv"),
    ).fillna("").sort_values(Partition.SORT_COLUMNS).reset_index(drop=True)

    ic(df_csv.iloc[:, [2, 3, 7]])

    df_parq = pd.read_parquet(
        cloudpathlib.AnyPath("dat/tiny.parq"),
        use_nullable_dtypes = True,
    ).fillna("").sort_values(Partition.SORT_COLUMNS).reset_index(drop=True)

    ic(df_parq.iloc[:, [2, 3, 7]])

    # general diff
    ic(df_csv.compare(df_parq))
    assert len(df_csv.compare(df_parq)) == 0

if __name__ == "__main__":
    test_pandas()


tests/test_csv_rdf.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Unit test coverage:

CSV => RDF

  * read a CSV file
  * construct a Partition internally
  * write an RDF file (TTL format)
"""

import tempfile

from icecream import ic
import cloudpathlib
import pyarrow.parquet as pq  # type: ignore
import pytest

from pynock import Partition


def test_parq_csv ():
    try:
        load_csv: str = "dat/tiny.csv"
        load_rdf: str = "dat/tiny.ttl"
        tmp_obs = tempfile.NamedTemporaryFile(mode="w+b", delete=True)

        # construct a Partition
        part: Partition = Partition(
            part_id = 0,
        )

        part.parse_rows(
            part.iter_load_csv(
                cloudpathlib.AnyPath(load_csv),
                encoding = "utf-8",
            ),
        )

        # write the partition as an RDF file
        part.save_file_rdf(
            cloudpathlib.AnyPath(tmp_obs.name),
            rdf_format = "ttl",
            encoding = "utf-8",
            sort = True,
        )

        # compare the respective texts
        obs_text: str = cloudpathlib.AnyPath(tmp_obs.name).read_text()
        exp_text: str = cloudpathlib.AnyPath(load_rdf).read_text()

        assert exp_text == obs_text

    except Exception as ex:
        ic(ex)

    finally:
        tmp_obs.close()


cli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Examples code for using `pynock`
"""

from icecream import ic  # type: ignore
import cloudpathlib
import pyarrow.parquet as pq  # type: ignore
import typer

from pynock import Partition

APP = typer.Typer()


@APP.command("load-parq")
def cli_load_parq (
    *,
    load_parq: str = typer.Option(..., "--file", "-f", help="input Parquet file"),
    save_csv: str = typer.Option(None, "--save-csv", help="output as CSV"),
    save_rdf: str = typer.Option(None, "--save-rdf", help="output as RDF"),
    rdf_format: str = typer.Option("ttl", "--format", help="RDF format: ttl, rdf, jsonld, etc."),
    encoding: str = typer.Option("utf-8", "--encoding", help="output encoding"),
    dump: bool = typer.Option(False, "--dump", help="dump the data, only"),
    sort: bool = typer.Option(False, "--sort", help="sort the output"),
    debug: bool = False,
    ) -> None:
    """
Load a Parquet file into a graph partition, optionally converting and
saving to different formats.
    """
    part: Partition = Partition(
        part_id = 0,
    )

    parq_file: pq.ParquetFile = pq.ParquetFile(load_parq)

    # in this case, only print what Parquet has parsed then quit
    if dump:
        part.dump_parquet(parq_file)
        return

    part.parse_rows(
        part.iter_load_parquet(
            parq_file,
            debug = debug,
        ),
        debug = debug,
    )

    if debug:
        ic(part)

    # next, handle the output options
    if save_csv is not None:
        part.save_file_csv(
            cloudpathlib.AnyPath(save_csv),
            encoding = encoding,
            sort = sort,
            debug = debug,
        )

    if save_rdf is not None:
        part.save_file_rdf(
            cloudpathlib.AnyPath(save_rdf),
            rdf_format = rdf_format,
            encoding = encoding,
            sort = sort,
            debug = debug,
        )


@APP.command("load-csv")
def cli_load_csv (
    *,
    load_csv: str = typer.Option(..., "--file", "-f", help="input CSV file"),
    save_parq: str = typer.Option(None, "--save-parq", help="output as Parquet"),
    save_rdf: str = typer.Option(None, "--save-rdf", help="output as RDF"),
    rdf_format: str = typer.Option("ttl", "--format", help="RDF format: ttl, rdf, jsonld, etc."),
    encoding: str = typer.Option("utf-8", "--encoding", help="output encoding"),
    sort: bool = typer.Option(False, "--sort", help="sort the output"),
    debug: bool = False,
    ) -> None:
    """
Load a CSV file into a graph partition, optionally converting and
saving to different formats.
    """
    part: Partition = Partition(
        part_id = 0,
    )

    part.parse_rows(
        part.iter_load_csv(
            cloudpathlib.AnyPath(load_csv),
            encoding = encoding,
            debug = debug,
        ),
        debug = debug,
    )

    if debug:
        ic(part)

    # next, handle the output options
    if save_parq is not None:
        part.save_file_parquet(
            cloudpathlib.AnyPath(save_parq),
            sort = sort,
            debug = debug,
        )

    if save_rdf is not None:
        part.save_file_rdf(
            cloudpathlib.AnyPath(save_rdf),
            rdf_format = rdf_format,
            encoding = encoding,
            sort = sort,
            debug = debug,
        )


@APP.command("load-rdf")
def cli_load_rdf (
    *,
    load_rdf: str = typer.Option(..., "--file", "-f", help="input RDF file"),
    rdf_format: str = typer.Option("ttl", "--format", help="RDF format: ttl, rdf, jsonld, etc."),
    save_parq: str = typer.Option(None, "--save-parq", help="output as Parquet"),
    save_csv: str = typer.Option(None, "--save-csv", help="output as CSV"),
    encoding: str = typer.Option("utf-8", "--encoding", help="output encoding"),
    sort: bool = typer.Option(False, "--sort", help="sort the output"),
    debug: bool = False,
    ) -> None:
    """
Load an RDF file into a graph partition, optionally converting and
saving to different formats.
    """
    part: Partition = Partition(
        part_id = 0,
    )

    part.parse_rows(
        part.iter_load_rdf(
            cloudpathlib.AnyPath(load_rdf),
            rdf_format = rdf_format,
            encoding = encoding,
            debug = debug,
        ),
    )

    if debug:
        ic(part)

    # next, handle the output options
    if save_parq is not None:
        part.save_file_parquet(
            cloudpathlib.AnyPath(save_parq),
            sort = sort,
            debug = debug,
        )

    if save_csv is not None:
        part.save_file_csv(
            cloudpathlib.AnyPath(save_csv),
            encoding = encoding,
            sort = sort,
            debug = debug,
        )


if __name__ == "__main__":
    APP()


example.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Examples code for using `pynock`
"""

from icecream import ic  # type: ignore
import cloudpathlib
import pyarrow.parquet as pq  # type: ignore
import typer

from pynock import Partition

APP = typer.Typer()


@APP.command("load-parq")
def cli_load_parq (
    *,
    load_parq: str = typer.Option(..., "--file", "-f", help="input Parquet file"),
    save_csv: str = typer.Option(None, "--save-csv", help="output as CSV"),
    debug: bool = False,
    ) -> None:
    """
Load a Parquet file into a graph partition, optionally converting and
saving to different formats.
    """
    parq_file: pq.ParquetFile = pq.ParquetFile(load_parq)

    if debug:
        ic(parq_file.metadata)
        ic(parq_file.schema)
        ic(type(parq_file.schema))

    part: Partition = Partition(
        part_id = 0,
    )

    part.parse_rows(part.iter_load_parquet(parq_file))

    if debug:
        ic(part)

    # next, handle the output options
    if save_csv is not None:
        part.save_file_csv(cloudpathlib.AnyPath(save_csv))


@APP.command("load-csv")
def cli_load_csv (
    *,
    load_csv: str = typer.Option(..., "--file", "-f", help="input CSV file"),
    save_parq: str = typer.Option(None, "--save-parq", help="output as Parquet"),
    debug: bool = False,
    ) -> None:
    """
Load a CSV file into a graph partition, optionally converting and
saving to different formats.
    """
    part: Partition = Partition(
        part_id = 0,
    )

    part.parse_rows(part.iter_load_csv(cloudpathlib.AnyPath(load_csv)))

    if debug:
        ic(part)

    # next, handle the output options
    if save_parq is not None:
        part.save_file_parquet(cloudpathlib.AnyPath(save_parq))


@APP.command("load-rdf")
def cli_load_rdf (
    *,
    load_rdf: str = typer.Option(..., "--file", "-f", help="input RDF file"),
    rdf_format: str = typer.Option("ttl", "--format", help="RDF format: ttl, rdf, jsonld, etc."),
    save_parq: str = typer.Option(None, "--save-parq", help="output as Parquet"),
    save_csv: str = typer.Option(None, "--save-csv", help="output as CSV"),
    debug: bool = False,
    ) -> None:
    """
Load an RDF file into a graph partition, optionally converting and
saving to different formats.
    """
    part: Partition = Partition(
        part_id = 0,
    )

    part.parse_rows(part.iter_load_rdf(cloudpathlib.AnyPath(load_rdf), rdf_format))

    if debug:
        ic(part)

    # next, handle the output options
    if save_parq is not None:
        part.save_file_parquet(cloudpathlib.AnyPath(save_parq))

    if save_csv is not None:
        part.save_file_csv(cloudpathlib.AnyPath(save_csv))


if __name__ == "__main__":
    APP()


nock/nock.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Example graph serialization using low-level Parquet read/write
efficiently in Python.
"""

import json
import typing

from icecream import ic  # type: ignore  # pylint: disable=E0401
from pydantic import BaseModel  # pylint: disable=E0401,E0611
import pyarrow.parquet as pq  # type: ignore  # pylint: disable=E0401
import pyarrow.lib  # type: ignore  # pylint: disable=E0401
from rich.progress import track  # pylint: disable=E0401


NOT_FOUND: int = -1


######################################################################
## nodes

class Node (BaseModel):  # pylint: disable=R0903
    """
Representing a node (entity) in the graph.
    """
    BASED_LOCAL: typing.ClassVar[int] = -1

    node_id: int = NOT_FOUND
    name: str = ""
    shadow: int = BASED_LOCAL
    is_rdf: bool = True
    label_set: typing.Set[str] = set()
    truth: float = 1.0
    prop_map: typing.Dict[str, str] = {}
    edge_map: typing.Dict[int, list] = {}


######################################################################
## edges

class Edge (BaseModel):  # pylint: disable=R0903
    """
Representing an edge (arc) in the graph.
    """
    BLANK_RELATION: typing.ClassVar[int] = 0

    rel: int = BLANK_RELATION
    node_id: int = NOT_FOUND
    truth: float = 1.0
    prop_map: typing.Dict[str, str] = {}


######################################################################
## partitions

class Partition (BaseModel):  # pylint: disable=R0903
    """
Representing a partition in the graph.
    """
    part_id: int = NOT_FOUND
    next_node: int = 0
    nodes: typing.Dict[int, Node] = {}
    node_names: typing.Dict[str, int] = {}
    edge_rels: typing.List[str] = [ "" ]


    def create_node (
        self,
        node_name: str,
        *,
        debug: bool = False,  # pylint: disable=W0613
        ) -> int:
        """
Create a node, looking up first to avoid duplicates.
        """
        node_id: int = NOT_FOUND

        if node_name in self.node_names:
            node_id = self.node_names[node_name]
        else:
            node_id = self.next_node
            self.node_names[node_name] = node_id
            self.next_node += 1

        return node_id


    @classmethod
    def load_props (
        cls,
        props: str,
        *,
        debug: bool = False,  # pylint: disable=W0613
        ) -> typing.Dict[str, str]:
        """
Load property pairs from a JSON string.
        """
        prop_map: typing.Dict[str, str] = {}

        if props not in ("null", ""):
            prop_map = json.loads(props)

        return prop_map


    def populate_node (
        self,
        dat: dict,
        *,
        debug: bool = False,  # pylint: disable=W0613
        ) -> Node:
        """
Populate a Node object from the given Parquet row data.
        """
        node: Node = Node(
            name = dat["src_name"],
            truth = dat["truth"],
            is_rdf = dat["is_rdf"],
            shadow = dat["shadow"],
            label_set = set(dat["labels"].split(",")),
            prop_map = self.load_props(dat["props"]),
        )

        # add this node to the global list
        node.node_id = self.create_node(node.name)
        self.nodes[node.node_id] = node

        return node


    def get_edge_rel (
        self,
        rel_name: str,
        *,
        create: bool = False,  # pylint: disable=W0613
        debug: bool = False,  # pylint: disable=W0613
        ) -> int:
        """
Lookup the integer index for the named edge relation.
        """
        if rel_name not in self.edge_rels:
            self.edge_rels.append(rel_name)

        return self.edge_rels.index(rel_name)


    def populate_edge (
        self,
        dat: dict,
        node: Node,
        *,
        debug: bool = False,  # pylint: disable=W0613
        ) -> Edge:
        """
Populate an Edge object from the given Parquet row data.
        """
        edge: Edge = Edge(
            rel = self.get_edge_rel(dat["rel_name"], create=True),
            truth = dat["truth"],
            node_id = self.create_node(dat["dst_name"]),
            prop_map = self.load_props(dat["props"]),
        )

        # add this edge to its src node
        if edge.rel not in node.edge_map:
            node.edge_map[edge.rel] = []

        node.edge_map[edge.rel].append(edge)

        return edge


    @classmethod
    def iter_rows (
        cls,
        row_group: pyarrow.lib.Table,  # pylint: disable=I1101
        *,
        debug: bool = False,  # pylint: disable=W0613
        ) -> typing.Iterable:
        """
Iterate through the rows in a Parquet row group.
        """
        for r_idx in range(row_group.num_rows):
            dat: dict = {}

            for c_idx in range(row_group.num_columns):
                try:
                    key: str = row_group.column_names[c_idx]
                    col: pyarrow.lib.ChunkedArray = row_group.column(c_idx)  # pylint: disable=I1101
                    val: typing.Any = col[r_idx]
                    dat[key] = val.as_py()
                except IndexError as ex:
                    ic(ex, r_idx, c_idx)
                    return

            if debug:
                print()
                ic(r_idx, dat)

            yield dat


    def load_rows (
        self,
        pq_file: pq.ParquetFile,
        *,
        debug: bool = False,
        ) -> None:
        """
Load and parse all of the Parquet rows into a graph partition.
        """
        for i in range(pq_file.num_row_groups):
            row_group: pyarrow.lib.Table = pq_file.read_row_group(i)  # pylint: disable=I1101

            for dat in track(self.iter_rows(row_group), description=f"row group {i}"):
                # have we reached a row which begins a new node?
                if dat["edge_id"] < 0:
                    node = self.populate_node(dat)

                    if debug:
                        print()
                        ic(node)

                # otherwise this row is an edge for the most recent node
                else:
                    assert dat["src_name"] == node.name
                    # 'edge_id': 2,

                    edge = self.populate_edge(dat, node)

                    if debug:
                        ic(edge)


tiny.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
A minimal example using `pynock` to construct a partition
programmatically, based on the graph described in `dat/tiny.rdf`
"""

import typing

from icecream import ic
import cloudpathlib

from pynock import Partition, Node, Edge


if __name__ == "__main__":
    # initialize a partition
    part: Partition = Partition(
        part_id = 0,
    )

    # lookup/create the src node for the recipe
    # NB: this node has properties, which RDF cannot query
    src_name: str = "https://www.food.com/recipe/327593"
    src_node: typing.Optional[Node] = part.lookup_node(src_name)

    if src_node is None:
        src_node = Node(
            node_id = part.create_node_name(src_name),
            name = src_name,
            is_rdf = True,
            label_set = set(["Recipe"]),
            prop_map = {
                "minutes": 8,
                "name": "anytime crepes"
            },
        )

        part.add_node(src_node)

    # lookup/create a dst node for the "Egg" ingredient
    dst_name: str = "http://purl.org/heals/ingredient/ChickenEgg"
    dst_node: typing.Optional[Node] = part.lookup_node(dst_name)

    if dst_node is None:
        dst_node = Node(
            node_id = part.create_node_name(dst_name),
            name = dst_name,
            is_rdf = True,
            label_set = set(["Ingredient"]),
        )

        part.add_node(dst_node)

    # define an edge connecting src => dst for this ingredient
    edge: Edge = Edge(
        rel = part.get_edge_rel("http://purl.org/heals/food/uses_ingredient", create=True),
        node_id = dst_node.node_id,
    )

    src_node.add_edge(edge)

    # define a dst node for the "Milk" ingredient
    dst_name = "http://purl.org/heals/ingredient/CowMilk"

    dst_node = Node(
        node_id = part.create_node_name(dst_name),
        name = dst_name,
        is_rdf = True,
        label_set = set(["Ingredient"]),
    )

    part.add_node(dst_node)

    # define an edge connecting src => dst for this ingredient
    edge = Edge(
        rel = part.get_edge_rel("http://purl.org/heals/food/uses_ingredient", create=True),
        node_id = dst_node.node_id,
    )

    src_node.add_edge(edge)

    # define a dst node for the "Flour" ingredient
    # NB: this node has properties, which RDF cannot query
    dst_name = "http://purl.org/heals/ingredient/WholeWheatFlour"
    dst_node = part.lookup_node(dst_name)

    if dst_node is None:
        dst_node = Node(
            node_id = part.create_node_name(dst_name),
            name = dst_name,
            is_rdf = True,
            label_set = set(["Ingredient"]),
            prop_map = {
                "vegan": True,
            },
        )

        part.add_node(dst_node)

    # define an edge connecting src => dst for this ingredient
    edge = Edge(
        rel = part.get_edge_rel("http://purl.org/heals/food/uses_ingredient", create=True),
        node_id = dst_node.node_id,
    )

    src_node.add_edge(edge)

    # define a dst node for the "wtm:Recipe" parent
    dst_name = "http://purl.org/heals/food/Recipe"
    dst_node = part.lookup_node(dst_name)

    if dst_node is None:
        dst_node = Node(
            node_id = part.create_node_name(dst_name),
            name = dst_name,
            is_rdf = True,
            label_set = set(["top-level"]),
        )

        part.add_node(dst_node)

    # define an edge connecting src => dst for this inheritance
    edge = Edge(
        rel = part.get_edge_rel("http://www.w3.org/1999/02/22-rdf-syntax-ns#type", create=True),
        node_id = dst_node.node_id,
    )

    src_node.add_edge(edge)

    # serialize this partition to multiple formats
    part.save_file_parquet(
        cloudpathlib.AnyPath("foo.parq"),
    )

    part.save_file_csv(
        cloudpathlib.AnyPath("foo.csv"),
        sort = True,
    )

    part.save_file_rdf(
        cloudpathlib.AnyPath("foo.rdf"),
        rdf_format = "ttl",
    )

    # check the files "foo.*" to see what was constructed programmatically
    # also, here's a dataframe representation
    df = part.to_df()
    ic(df.head())

